src/Spice86.Core/Emulator/InterruptHandlers/Dos/Ems/ExpandedMemoryManager.cs

```diff
@@ -0,0 +1,1236 @@
+        if (_loggerService.IsEnabled(LogEventLevel.Debug))
+            _loggerService.Debug("EMS: {@MethodName}: {@Result}", nameof(GetStatus), _state.AH);
+    }
+    
+    public void GetPageFrameSegment() {
+        // Return page frame segment in BX.
+        _state.BX = EmmPageFrameSegment;
+        // Set good status.
+        _state.AH = EmmStatus.EmmNoError;
+        if (_loggerService.IsEnabled(LogEventLevel.Debug))
+            _loggerService.Debug("EMS: {@MethodName}: 0x{Result:X4}", nameof(GetPageFrameSegment), _state.BX);
+    }
+    
+    public void GetNumberOfPages() {
+        // Return total number of pages in DX.
+        _state.DX = (ushort) (TotalPages / 4);
+        // Return number of pages available in BX.
+        _state.BX = GetFreePages();
+        // Set good status.
+        _state.AH = EmmStatus.EmmNoError;
+        if (_loggerService.IsEnabled(LogEventLevel.Debug))
+            _loggerService.Debug("EMS: {@MethodName}: Total: {@Total} Free: {@Free}", nameof(GetNumberOfPages), _state.DX, _state.BX);
+    }
+    
+    /// <summary>
+    /// Allocates pages for a new handle.
+    /// </summary>
+    public void GetHandleAndAllocatePages() {
+        ushort handles = _state.DX;
+        ushort pages = _state.BX;
+        if (_loggerService.IsEnabled(LogEventLevel.Debug))
+            _loggerService.Debug("EMS: {@MethodName}: requesting {@Pages} pages", nameof(GetHandleAndAllocatePages), pages);
+        _state.AH = EmmAllocateMemory(pages, ref handles, false);
+        _state.DX = handles;
+        if (_state.AH == EmmStatus.EmmNoError && _loggerService.IsEnabled(LogEventLevel.Debug))
+            _loggerService.Debug("EMS: {@MethodName}: got handle {@Handle} [{HandleName}]", nameof(GetHandleAndAllocatePages), handles, GetHandleName(handles));
+    }
+    
+    /// <summary>
+    /// Maps or unmaps a physical page.
+    /// </summary>
+    public void MapExpandedMemoryPage() {
+        ushort handle = _state.DX;
+        ushort physicalPage = _state.AL;
+        ushort logicalPage = _state.BX;
+        _state.AH = EmmMapPage(physicalPage, handle, logicalPage);
+        _state.DX = handle;
+    }
+
+    /// <summary>
+    /// Maps data from a logical page to a physical page.
+    /// </summary>
+    /// <param name="physicalPage">Physical page id.</param>
+    /// <param name="handle">The assigned handle id.</param>
+    /// <param name="logicalPage">The logical page id.</param>
+    private void RegisterEmmMapping(int physicalPage, ushort handle, ushort logicalPage) {
+        uint destAddress = MemoryUtils.ToPhysicalAddress(EmmPageFrameSegment, (ushort) (EmmPageSize * physicalPage));
+        EmmMapping mapping = EmmMappings[physicalPage];
+        if (_loggerService.IsEnabled(LogEventLevel.Debug)) {
+            _loggerService.Debug("EMS: {@MethodName}: Map of {@EmmMapping}",
+                nameof(RegisterEmmMapping), mapping);
+        }
+        _memory.RegisterMapping(destAddress, EmmPageSize, this);
```

Comment:
You don't want to register a mapping from our address space to a physical page every time you create a mapping between a physical page and a logical page.
You can map the whole page frame at once in the constructor (EmmPageFrameSegment to physicalPageCount (4) * EmmPageSize (16k))